lib/ansible/plugins/action/package.py

```diff
@@ -38,31 +39,46 @@ def run(self, tmp=None, task_vars=None):
         self._supports_async = True
 
         result = super(ActionModule, self).run(tmp, task_vars)
-        del tmp  # tmp no longer has any effect
 
         module = self._task.args.get('use', 'auto')
 
-        if module == 'auto':
-            try:
-                if self._task.delegate_to:  # if we delegate, we should use delegated host's facts
-                    module = self._templar.template("{{hostvars['%s']['ansible_facts']['pkg_mgr']}}" % self._task.delegate_to)
-                else:
-                    module = self._templar.template('{{ansible_facts.pkg_mgr}}')
-            except Exception:
-                pass  # could not get it from template!
-
         try:
             if module == 'auto':
-                facts = self._execute_module(
-                    module_name='ansible.legacy.setup',
-                    module_args=dict(filter='ansible_pkg_mgr', gather_subset='!all'),
-                    task_vars=task_vars)
-                display.debug("Facts %s" % facts)
-                module = facts.get('ansible_facts', {}).get('ansible_pkg_mgr', 'auto')
-
-            if module != 'auto':
+
+                if self._task.delegate_to:
+                    hosts_vars = task_vars['hostvars'][self._task.delegate_to]
+                    tvars = combine_vars(self._task.vars, task_vars.get('delegated_vars', {}))
+                else:
+                    hosts_vars = task_vars
+                    tvars = task_vars
+
+                # use config
+                module = tvars.get('ansible_package_use', None)
+
+                if not module:
+                    # no use, no config, get from facts
+                    if hosts_vars.get('_ansible_facts_gathered', False):
```

Comment:
I don't think this will be sufficient.  It's set by `gather_facts` but there is nothing saying that `gather_subset` may have excluded `pkg_mgr`.

`gather_subset: ['!min']` for example would prevent it from being gathered.

Reply:
Going back to `try/except KeyError` may be best

Reply:
made it conditional on the actual data